soroban-env-host/src/events/internal.rs

```diff
@@ -0,0 +1,109 @@
+use super::{DebugEvent, Events, HostEvent};
+use crate::{
+    budget::Budget,
+    host::metered_clone::MeteredClone,
+    xdr,
+    xdr::{Hash, ScObject},
+    Host, HostError, MeteredVector, Object, RawVal,
+};
+use log::debug;
+
+#[derive(Clone, Debug)]
+pub(crate) struct InternalContractEvent {
+    pub(crate) type_: xdr::ContractEventType,
+    pub(crate) contract_id: Option<Hash>,
+    pub(crate) topics: Object,
+    pub(crate) data: RawVal,
+}
+
+impl InternalContractEvent {
+    // Metering: covered by components
+    pub fn to_xdr(self, host: &Host) -> Result<xdr::ContractEvent, HostError> {
+        let topics = if let ScObject::Vec(v) = host.from_host_obj(self.topics)? {
+            Ok(v)
+        } else {
+            Err(host.err_status(xdr::ScHostObjErrorCode::UnexpectedType))
+        }?;
+        let data = host.from_host_val(self.data)?;
+        Ok(xdr::ContractEvent {
+            ext: xdr::ExtensionPoint::V0,
+            contract_id: self.contract_id,
+            type_: self.type_,
+            body: xdr::ContractEventBody::V0(xdr::ContractEventV0 { topics, data }),
+        })
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub(crate) enum InternalEvent {
+    Contract(InternalContractEvent),
+    Debug(DebugEvent),
+    #[default]
+    None,
+}
+
+impl MeteredClone for InternalEvent {}
+
+#[derive(Clone, Default)]
+pub(crate) struct InternalEventsBuffer {
+    pub(crate) vec: MeteredVector<InternalEvent>,
+}
+
+impl InternalEventsBuffer {
+    // Records an InternalEvent
+    // Metering covered by the `MeteredVec`.
+    pub fn record(&mut self, e: InternalEvent, budget: &Budget) -> Result<(), HostError> {
+        self.vec = self.vec.push_back(e, budget)?;
+        Ok(())
+    }
+
+    // Metering covered by the `MeteredVec`
+    pub fn rollback(&mut self, events: usize, budget: &Budget) -> Result<(), HostError> {
+        let mut i = events;
+        let mut v = self.vec.clone();
+        loop {
+            match v.get(i, budget)? {
```

Comment:
Might be nice to do this in a `retain()` or `filter()`-like call that builds the new vector in a single pass.

Reply:
so something like
```
v.iter_mut().enumerate().retain_mut(|(i, event) {
  if i < events {
    if let ContractEvent(_) = *e {
      *e = DebugEvent("rolled back contractevent");
    }
    true
  } else {
     false
  }
})
```

Reply:
I have added a `MeteredVec::retain_mut` and updated here to use it.